### PR TITLE
Feature/#313 autosave local storage step1

### LINKED ISSUE #313: Autosave local storage, step-1
# Autosave using LocalStorage

We are planning to implement an autosave feature to locally store the `canvasSchema`, with the goal of saving it every minute and retrieve it when the application is loaded.

## Implementation Plan

1. **Create Autosave Folder**: Create a new folder named `autosave` within the `core` directory.

2. **Autosave Hook**: Develop a custom hook named `useAutosave`, which will handle the autosave functionality.

3. **Business Logic**:  May require a `autose.bussines.ts` file to store functions related to autosave operations.

### Details

- The autosave feature will save our canvas in the following format:
"autoSaveFile": { filename: <filename>, canvasSchema: <canvasSchema> }
The filename can be either undefined or have a specific value, which will be obtained from the filename saved in the `CanvasSettingsProvider`.

- Errors: If the autosave process encounters an error, it should not disrupt the application. Instead, the application should continue functioning as usual, and save the document 1 minute later. Upon detecting an error during file recovery from local storage, we will delete the local file and create a new canvasSchema.

### Proposed Hook Example

Here's an example of the `useAutosave` hook that we have created:

```typescript
import { useEffect } from 'react';
import { saveToLocal } from './localStorageUtils';

interface CanvasSchema {
// Define the structure of your canvasSchema
}

interface Options {
interval?: number;
}

const useAutosave = () => {
 const { canvasSchema } = useCanvasSchemaContext();
 const { filename} = useCanvasViewSettingsContext();

const  interval = 60000  // Default interval is 1 minute
const timerRef = useRef<NodeJS.Timeout | null>(null);
const [isSaving, setIsSaving] = useState(false);

useEffect(() => {
  const autosaveHandler = () => {
    if (!isSaving) {
      setIsSaving(true);
      saveToLocal('autoSaveFile', {
        filename: /* Obtain the filename */,
        canvasSchema: canvasSchema
      });
      setIsSaving(false);
    }
  };

  timerRef.current = setInterval(autosaveHandler, interval);

  return () => {
    if (timerRef.current) {
      clearInterval(timerRef.current);
      timerRef.current = null;
    }
  };
}, [canvasSchema, interval, isSaving]);

const stopAutosave = () => {
  if (timerRef.current) {
    clearInterval(timerRef.current);
    timerRef.current = null;
  }
};

return stopAutosave;
};

export default useAutosave;
```

Feel free to adjust the implementation according to your specific requirements and preferences.

### PR DESCRIPTION
**Pull request by AlbertoSSC**
Closed #313

### CODE REVIEW

File: src/core/autosave/autosave.Hook.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { useEffect, useRef, useState } from 'react';
```
Comment (brauliodiez) [inline src/core/autosave/autosave.Hook.tsx:1]: Filename should be

`autosave.hook.ts`

(lower case)

File: src/core/autosave/autosave.Hook.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { saveToLocal, retrieveFromLocal } from './autosave.business';
+
+const useAutosave = () => {
+  const { canvasSchema, setCanvasSchema } = useCanvasSchemaContext();
+  const { filename, setLoadSample } = useCanvasViewSettingsContext();
+
+  const INTERVAL = 6000;
+  const timerRef = useRef<NodeJS.Timeout | null>(null);
+
+  const [isSaving, setIsSaving] = useState<boolean>(false);
```
Comment (brauliodiez) [inline src/core/autosave/autosave.Hook.tsx:17]: Mmm... not sure if we need _isSaving_ on the one hand javaScript is singlethreaded by default, in the other save to local storage not sure if it's blocking or not (I/O)

File: src/core/autosave/autosave.Hook.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  useCanvasViewSettingsContext,
+} from '@/core/providers';
+
+import { saveToLocal, retrieveFromLocal } from './autosave.business';
+
+const useAutosave = () => {
+  const { canvasSchema, setCanvasSchema } = useCanvasSchemaContext();
+  const { filename, setLoadSample } = useCanvasViewSettingsContext();
+
+  const INTERVAL = 6000;
```
[the remark this answers is not in the document]
Comment (brauliodiez) [reply src/core/autosave/autosave.Hook.tsx:14]: Interface should be at least 60 seconds (one minute)

File: src/core/autosave/autosave.Hook.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  }, [canvasSchema, filename, isSaving]);
+
+  const stopAutosave = () => {
+    if (timerRef.current) {
+      clearInterval(timerRef.current);
+      timerRef.current = null;
+    }
+  };
+
+  return { stopAutosave };
```
Comment (brauliodiez) [inline src/core/autosave/autosave.Hook.tsx:55]: Oooops... thinking about edge case what happens if we have two tabs opened with two different models? :-)

File: src/core/autosave/autosave.business.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+export const saveToLocal = (key: string, value: CanvasSchema) => {
+  try {
+    const deepCopyValue = JSON.parse(JSON.stringify(value));
+    const stringifiedDeepCopyValue = JSON.stringify(deepCopyValue);
+    localStorage.setItem(key, stringifiedDeepCopyValue);
+  } catch (error) {
+    console.error('Autosave error', error);
+    setTimeout(() => {
+      saveToLocal(key, value);
+    }, 30000);
```
Comment (brauliodiez) [inline src/core/autosave/autosave.business.ts:20]: I think we should not call this in the catch, if something went wrong, well on next interval will try again

File: src/core/autosave/autosave.business.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+export interface CanvasSchema {
+  filename: string | undefined;
+  canvasSchema: DatabaseSchemaVm;
+}
+
+export const saveToLocal = (key: string, value: CanvasSchema) => {
+  try {
+    const deepCopyValue = JSON.parse(JSON.stringify(value));
+    const stringifiedDeepCopyValue = JSON.stringify(deepCopyValue);
```
Comment (brauliodiez) [inline src/core/autosave/autosave.business.ts:14]: Why is not enough with JSON.stringify?

File: src/core/providers/canvas-schema/canvas.const.ts [hunk 7]
```diff
@@ -11,6 +11,8 @@ const ROW_HEIGHT = FONT_SIZE + ROW_PADDING; // Height of the table line
 const HORIZONTAL_LEFT_EXTENSION = 20; // Horizontal extension for the relation line 1:1
 const HEADER_TITLE_GAP = 15;
 const CANVAS_PADDING = 100; // Represents the additional space around the canvas, we use it to add a space for the tables when we export the schema
+const MAX_PLACEMENT_ATTEMPTS = 20; // Maximum number of attempts to place a table without overlapping
+const TABLE_SHIFT_DISTANCE = 30; // Distance to shift the table when it overlaps with another table
```
Comment (brauliodiez) [inline src/core/providers/canvas-schema/canvas.const.ts:15]: There's something strange with this updates, they don't belong to the case, maybe you need to update from latest vnext cut?

File: src/pods/canvas/canvas.pod.tsx [hunk 8]
```diff
@@ -18,8 +18,15 @@ import {
...
 import { EditRelationPod } from '../edit-relation';
 import { mFlix } from './m-flix.mock.data';
+import useAutosave from '@/core/autosave/autosave.hook';
 
 export const CanvasPod: React.FC = () => {
+  const { stopAutosave } = useAutosave();
+
+  React.useEffect(() => {
+    return stopAutosave;
+  }, []);
```
Comment (brauliodiez) [inline src/pods/canvas/canvas.pod.tsx:28]: Not sure if we can even place this at app level

Comment (brauliodiez) [reply src/pods/canvas/canvas.pod.tsx:28]: Maybe we should expose a

`startAutosave` and `stopAutosave`

So it could be something like

```diff
React.useEffect(() => {
  startAutosave;

  return () => stopAutosave();
}, []);
```

File: src/pods/toolbar/components/export-button/export-button.business.ts [hunk 9]
```diff
@@ -28,10 +29,68 @@ export const calculateTableEndYPosition = (table: TableVm): number => {
   return table.y + tableHeight;
 };
 
+export const doesTableOverlap = (
```
Comment (brauliodiez) [inline src/pods/toolbar/components/export-button/export-button.business.ts:32]: Same thing here this code is not related to the case, and likely you are stuck on older version of vnext?

File: src/core/autosave/autosave.business.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      saveToLocal(key, value);
+    }, 30000);
+  }
+};
+
+export const retrieveFromLocal = (
+  setLoadSample: (T: boolean) => void
+): DatabaseSchemaVm => {
+  try {
+    const value = localStorage.getItem('autoSaveFile');
```
Comment (brauliodiez) [inline src/core/autosave/autosave.business.ts:28]: Check if we can use the loca storage commo wrapper, if not expand functionallity on our common stuff

**Comment by AlbertoSSC** [comment]
closed to do a new one, with vnext updated
